firedrake/matrix.py

```diff
@@ -153,3 +170,42 @@ def assemble(self):
         # Ensures that if the matrix changed, the preconditioner is
         # updated if necessary.
         self.petscmat.assemble()
+
+
+class AssembledMatrix(MatrixBase):
+    """A representation of an matrix without knowing the underlying form.
+     This class wraps the relevant information for Python PETSc matrix.
+
+    :arg a: A tuple of the arguments the matrix represents
+
+    :arg bcs: an iterable of boundary conditions to apply to this
+        :class:`Matrix`.  May be `None` if there are no boundary
+        conditions to apply.
+
+    :arg petscmat: the already constructed petsc matrix this object represents.
+
+
+    .. note::
+
+        This object acts to the right on an assembled :class:`.Function`
+        and to the left on an assembled cofunction (currently represented
+        by a :class:`.Function`).
+
+    """
+    def __init__(self, a, bcs, petscmat, *args, **kwargs):
+        super(AssembledMatrix, self).__init__(a, bcs, "assembled")
+
+        self.petscmat = petscmat
+
+        # this allows call to self.M.handle without a new class
+        self.M = SimpleNamespace(handle=self.mat())
```

Comment:
I _think_ it should be the case that we should expunge `matrix.M.handle` and just use `matrix.petscmat` everywhere, but maybe this is too much for this PR.

Reply:
If necessary, just use `self.M = SimpleNamespace(handle=self.petscmat)` and remove the need for the `mat()` method?